### PR TITLE
Add support for decrypted NCA files

### PR DESCRIPTION
**Pull request by DarkLordZach**
IMPORTANT: Must be decrypted before hand. Use `hactool --titlekey=<titlekey here> --plaintext=<out file> <in file` to decrypt an nca.

This *should* work. I have tested a bit but I only have 3 games to test (mk8dx, botw, and kingdom battle) all worked, let me know if you encounter issues on another game.

So this adds:
- Ability to pick nca files in 'Load File'
- Scans for nca files in game list
- Will show encrypted nca files in list as Unknown type.
- Runs games in nca format

Big thanks to @shinyquagsire23 for helping out.

### CODE REVIEW

File: src/core/loader/nca.cpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    for (int i = 0; i < pfs.size(); ++i) {
+        if (IsPfsExeFs(pfs[i]))
+            return i;
+    }
+
+    return -1;
+}
+
+u64 Nca::GetExeFsFileOffset(const std::string& file_name) {
+    if (GetExeFsPfsId() == 255)
```
[the remark this answers is not in the document]
Comment (DarkLordZach) [reply src/core/loader/nca.cpp]: I looked into it (a bit) and if theres no exefs in the nca (say its a icon or manual or whatever not a pgm) the GetExeFsFile will return an empty vector and the Load method will return an error as it tries to parse said empty vector as a npdm file.

File: src/core/loader/nca.cpp [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+#include "core/hle/kernel/process.h"
+#include "core/hle/kernel/resource_limit.h"
+#include "core/hle/service/filesystem/filesystem.h"
+#include "core/loader/nca.h"
+#include "core/loader/nso.h"
+#include "core/memory.h"
+
+namespace Loader {
+
+enum NcaContentType {
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: `enum class NcaContentType : u8` ?

File: src/core/loader/nca.cpp [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    INSERT_PADDING_BYTES(0x8);
+};
+static_assert(sizeof(NcaSectionTableEntry) == 0x10, "NcaSectionTableEntry has incorrect size.");
+
+struct NcaHeader {
+    u8 rsa_signature_1[0x100];
+    u8 rsa_signature_2[0x100];
+    u32_le magic;
+    u8 is_system;
+    u8 content_type;
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: `NcaContentType`?

File: src/core/loader/nca.cpp [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+struct NcaSectionTableEntry {
+    u32_le media_offset;
+    u32_le media_end_offset;
+    INSERT_PADDING_BYTES(0x8);
+};
+static_assert(sizeof(NcaSectionTableEntry) == 0x10, "NcaSectionTableEntry has incorrect size.");
+
+struct NcaHeader {
+    u8 rsa_signature_1[0x100];
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: Prefer `std::array` to C-style arrays

Comment (bunnei) [reply src/core/loader/nca.cpp]: (applies to all new arrays with this change)

File: src/core/loader/nso.cpp [hunk 5]
```diff
@@ -77,22 +91,64 @@ static std::vector<u8> ReadSegment(FileUtil::IOFile& file, const NsoSegmentHeade
...
 static constexpr u32 PageAlignSize(u32 size) {
     return (size + Memory::PAGE_MASK) & ~Memory::PAGE_MASK;
 }
 
+VAddr AppLoader_NSO::LoadModule(const std::string& name, const std::vector<u8>& file_data,
+                                VAddr load_base) {
+    if (file_data.size() < sizeof(NsoHeader))
+        return {};
+
+    const NsoHeader nso_header = *reinterpret_cast<const NsoHeader* const>(file_data.data());
```
Comment (bunnei) [inline src/core/loader/nso.cpp]: `std::memcpy`?

File: src/core/loader/nca.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+#include "core/file_sys/partition_filesystem.h"
+#include "core/file_sys/program_metadata.h"
+#include "core/hle/kernel/kernel.h"
+#include "core/loader/linker.h"
+#include "core/loader/loader.h"
+
+namespace Loader {
+
+// TODO(DarkLordZach): Add support for encrypted.
+struct Nca {
```
Comment (bunnei) [inline src/core/loader/nca.h]: nit: `class Nca final` ?

Comment (bunnei) [reply src/core/loader/nca.h]: Also, this can be moved to the .cpp, and just forward declared here

File: src/core/loader/nca.h [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    u64 romfs_size = 0;
+
+    bool valid = false;
+
+    FileUtil::IOFile file;
+    std::string path;
+};
+
+/// Loads an NCA file
+class AppLoader_NCA final : public AppLoader, Linker {
```
Comment (bunnei) [inline src/core/loader/nca.h]: I'm not sure we need this `Linker` stuff, right? Remove if unused (and remove header)

File: src/core/loader/nca.h [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+#include "core/loader/linker.h"
+#include "core/loader/loader.h"
+
+namespace Loader {
+
+// TODO(DarkLordZach): Add support for encrypted.
+struct Nca {
+    Nca(FileUtil::IOFile&& file, std::string path);
+
+    FileSys::PartitionFilesystem GetPfs(u8 id);
```
Comment (bunnei) [inline src/core/loader/nca.h]: Can be a `const` method

File: src/core/loader/nca.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    boost::optional<u8> GetExeFsPfsId();
+
+    u64 GetExeFsFileOffset(const std::string& file_name);
+    u64 GetExeFsFileSize(const std::string& file_name);
+
+    u64 GetRomFsOffset();
+    u64 GetRomFsSize();
+
+    bool IsValid();
```
Comment (bunnei) [inline src/core/loader/nca.h]: Cursory look, all of the above methods can also be `const`

File: src/core/loader/nca.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace Loader {
+
+// TODO(DarkLordZach): Add support for encrypted.
+struct Nca {
+    Nca(FileUtil::IOFile&& file, std::string path);
+
+    FileSys::PartitionFilesystem GetPfs(u8 id);
+
+    boost::optional<u8> GetExeFsPfsId();
```
Comment (bunnei) [inline src/core/loader/nca.h]: `private`? Ditto to any other functions that do not need to be used outside the public API for this class

File: src/core/loader/nca.cpp [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+Nca::Nca(FileUtil::IOFile&& in_file, std::string path) : file(std::move(in_file)), path(path) {
+    file.Seek(0, SEEK_SET);
+    std::array<u8, 0x400> header_array{};
+    if (0x400 != file.ReadBytes(header_array.data(), 0x400))
+        NGLOG_CRITICAL(Loader, "File reader errored out during header read.");
+
+    NcaHeader header = *reinterpret_cast<NcaHeader*>(header_array.data());
+    valid = IsValidNca(header);
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: Should we abort if not valid?

File: src/core/loader/nca.cpp [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+FileType AppLoader_NCA::IdentifyType(FileUtil::IOFile& file, const std::string&) {
+    file.Seek(0, SEEK_SET);
+    std::array<u8, 0x400> header_enc_array{};
+    if (0x400 != file.ReadBytes(header_enc_array.data(), 0x400))
+        return FileType::Error;
+
+    // NcaHeader header = DecryptHeader(header_enc_array);
+    // TODO(DarkLordZach): Assuming everything is decrypted. Add crypto support.
+    NcaHeader header = *reinterpret_cast<NcaHeader*>(header_enc_array.data());
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: `std::memcpy`?

File: src/core/loader/nca.cpp [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+           header.magic == Common::MakeMagic('N', 'C', 'A', '3');
+}
+
+Nca::Nca(FileUtil::IOFile&& in_file, std::string path) : file(std::move(in_file)), path(path) {
+    file.Seek(0, SEEK_SET);
+    std::array<u8, 0x400> header_array{};
+    if (0x400 != file.ReadBytes(header_array.data(), 0x400))
+        NGLOG_CRITICAL(Loader, "File reader errored out during header read.");
+
+    NcaHeader header = *reinterpret_cast<NcaHeader*>(header_array.data());
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: `std::memcpy`?

File: src/core/loader/nca.cpp [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+static_assert(sizeof(Pfs0Superblock) == 0x200, "Pfs0Superblock has incorrect size.");
+
+static bool IsValidNca(const NcaHeader& header) {
+    return header.magic == Common::MakeMagic('N', 'C', 'A', '2') ||
+           header.magic == Common::MakeMagic('N', 'C', 'A', '3');
+}
+
+Nca::Nca(FileUtil::IOFile&& in_file, std::string path) : file(std::move(in_file)), path(path) {
+    file.Seek(0, SEEK_SET);
+    std::array<u8, 0x400> header_array{};
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: Replace all usages of `0x400` with `sizeof(NcaHeader)`

File: src/core/loader/nca.cpp [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+        NcaSectionHeaderBlock block = *reinterpret_cast<NcaSectionHeaderBlock*>(array.data());
+
+        if (block.filesystem_type == NSFS_ROMFS) {
+            romfs_offset = (header.section_tables[i].media_offset << 9);
+            romfs_size = (header.section_tables[i].media_end_offset << 9) - romfs_offset;
+        } else if (block.filesystem_type == NSFS_PFS0) {
+            Pfs0Superblock sb;
+            file.Seek(0x400 + (i << 9), SEEK_SET);
+            if (0x200 != file.ReadBytes(&sb, 0x200))
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: `sizeof(Pfs0Superblock)` instead of `0x200`?

File: src/core/loader/nca.cpp [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+    valid = IsValidNca(header);
+
+    int number_sections =
+        std::count_if(std::begin(header.section_tables), std::end(header.section_tables),
+                      [](NcaSectionTableEntry entry) { return entry.media_offset > 0; });
+
+    for (int i = 0; i < number_sections; ++i) {
+        file.Seek(0x400 + (i << 9), SEEK_SET);
+        std::array<u8, 0x10> array{};
+        if (0x10 != file.ReadBytes(array.data(), 0x10))
```
Comment (bunnei) [inline src/core/loader/nca.cpp]: Replace all `0x10`'s with `sizeof(NcaSectionTableEntry)`

File: src/core/loader/nca.cpp [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+
+    NcaHeader header = *reinterpret_cast<NcaHeader*>(header_array.data());
+    valid = IsValidNca(header);
+
+    int number_sections =
+        std::count_if(std::begin(header.section_tables), std::end(header.section_tables),
+                      [](NcaSectionTableEntry entry) { return entry.media_offset > 0; });
+
+    for (int i = 0; i < number_sections; ++i) {
+        file.Seek(0x400 + (i << 9), SEEK_SET);
```
Comment (Subv) [inline src/core/loader/nca.cpp]: What does this i << 9 mean? Please make it clearer with either named constants or a comment (preferably both)

File: src/core/loader/nca.cpp [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    for (size_t i = 0; i < pfs.size(); ++i) {
+        if (IsPfsExeFs(pfs[i]))
+            return i;
+    }
+
+    return boost::none;
+}
+
+u64 Nca::GetExeFsFileOffset(const std::string& file_name) {
+    if (GetExeFsPfsId() == boost::none)
```
Comment (Subv) [inline src/core/loader/nca.cpp]: Call GetExeFsPfsId once and store the result instead of calling it three times.

File: src/core/loader/nca.cpp [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+        return 0;
+    return pfs[*GetExeFsPfsId()].GetFileOffset(file_name) + pfs_offset[*GetExeFsPfsId()];
+}
+
+u64 Nca::GetExeFsFileSize(const std::string& file_name) {
+    if (GetExeFsPfsId() == boost::none)
+        return 0;
+    return pfs[*GetExeFsPfsId()].GetFileSize(file_name);
```
Comment (Subv) [inline src/core/loader/nca.cpp]: same

File: src/core/loader/nca.cpp [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+u64 Nca::GetRomFsSize() {
+    return romfs_size;
+}
+
+bool Nca::IsValid() {
+    return valid;
+}
+
+std::vector<u8> Nca::GetExeFsFile(const std::string& file_name) {
+    std::vector<u8> out(GetExeFsFileSize(file_name));
```
Comment (Subv) [inline src/core/loader/nca.cpp]: Use variables instead of calling the same function multiple times

File: src/core/loader/nca.cpp [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+AppLoader_NCA::AppLoader_NCA(FileUtil::IOFile&& file, std::string filepath)
+    : AppLoader(std::move(file)), filepath(std::move(filepath)) {}
+
+FileType AppLoader_NCA::IdentifyType(FileUtil::IOFile& file, const std::string&) {
+    file.Seek(0, SEEK_SET);
+    std::array<u8, 0x400> header_enc_array{};
+    if (0x400 != file.ReadBytes(header_enc_array.data(), 0x400))
+        return FileType::Error;
+
+    // NcaHeader header = DecryptHeader(header_enc_array);
```
Comment (Subv) [inline src/core/loader/nca.cpp]: Remove this commented out code

**Pull request merged**